src/tasks/WasmAppBuilder/IcallTableGenerator.cs

```diff
@@ -0,0 +1,301 @@
+        foreach (var aname in assemblies)
+        {
+            var a = mlc.LoadFromAssemblyPath(aname);
+            foreach (var type in a.GetTypes())
+                ProcessType(type);
+        }
+
+        using (var w = File.CreateText(OutputPath!))
+            EmitTable (w);
+    }
+
+    private void EmitTable (StreamWriter w)
+    {
+        var assemblyMap = new Dictionary<string, string> ();
+        foreach (var icall in _icalls)
+            assemblyMap [icall.Assembly!] = icall.Assembly!;
+
+        foreach (var assembly in assemblyMap.Keys)
+        {
+            var sorted = _icalls.Where (i => i.Assembly == assembly).ToArray ();
+            Array.Sort (sorted);
+
+            string aname;
+            if (assembly == "System.Private.CoreLib")
+                aname = "corlib";
+            else
+                aname = assembly.Replace (".", "_");
+            w.WriteLine ($"#define ICALL_TABLE_{aname} 1\n");
+
+            w.WriteLine ($"static int {aname}_icall_indexes [] = {{");
+            foreach (var icall in sorted)
+                w.WriteLine (string.Format ("{0},", icall.TokenIndex));
+            w.WriteLine ("};");
+            foreach (var icall in sorted)
+                w.WriteLine (GenIcallDecl (icall));
+            w.WriteLine ($"static void *{aname}_icall_funcs [] = {{");
+            foreach (var icall in sorted)
+            {
+                w.WriteLine (string.Format ("// token {0},", icall.TokenIndex));
+                w.WriteLine (string.Format ("{0},", icall.Func));
+            }
+            w.WriteLine ("};");
+            w.WriteLine ($"static uint8_t {aname}_icall_handles [] = {{");
+            foreach (var icall in sorted)
+                w.WriteLine (string.Format ("{0},", icall.Handles ? "1" : "0"));
+            w.WriteLine ("};");
+        }
+    }
+
+    // Read the icall table generated by mono --print-icall-table
+    private void ReadTable (string filename)
+    {
+        JsonDocument json;
+        using (var stream = File.Open (filename, FileMode.Open))
+        {
+            json = JsonDocument.Parse (stream);
+        }
+
+        var arr = json.RootElement;
+        foreach (var v in arr.EnumerateArray ())
+        {
+            var className = v.GetProperty ("klass").GetString ()!;
+            if (className == "")
+                // Dummy value
+                continue;
+            var icallClass = new IcallClass (className);
+            _runtimeIcalls [icallClass.Name] = icallClass;
+            foreach (var icall_j in v.GetProperty ("icalls").EnumerateArray ())
+            {
+                if (!icall_j.TryGetProperty ("name", out var nameElem))
+                    continue;
+                string name = nameElem.GetString ()!;
+                string func = icall_j.GetProperty ("func").GetString ()!;
+                bool handles = icall_j.GetProperty ("handles").GetBoolean ();
+
+                icallClass.Icalls [name] = new Icall (name, func, handles);
```

Comment:
if we have duplicate items with the same 'name' this will silently overwrite the preceding ones with the new one, I think it might be better to use .Add instead so it will throw a duplicate key exception